### ISSUE TITLE
Typescript does not automatically infer existence & validity of a variable 

### ISSUE THREAD

**Issue by aghArdeshir** [microsoft/TypeScript#45194]
# Bug Report
When I want to check existence of a variable and then do something with it, I have to do it right before the usage, I cannot move it somewhere else and this makes refactoring hard/impossible

### 🔎 Search Terms
infer, infer automatically

### 🕗 Version & Regression Information
Typescript: 4.3.5

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABDASgUwmmA3NA5EAWwCM0AnARgAoBDALkTCNLIBpFiGmTyBKRAN4AoRKMQQEAZzgAbNADoZcAOa1EAag68A3EIC+Q0JFgJk6TDnzNyAJlpdrbDg55l+wseKmyFS1TQ0tXQMjaHgkVAwsXAJXAGZ7Rkd2TiTXdxExCTBpOUUVNU1iHX1DcDDTSIsYxwAWRO4WFJcWDM9s3N8CgKKSgzLjcMQABzI4ABMQTFiWKjaxMjQoEDIkAFkaKAALeTIaMHG4QjnEAD5EAAZ5AFZEAH5ECkQGcHG0YBgwNHHgoSEAegAVICRIDEABRAAeNEIwzkj0Q202iAA7nAyABrSQpEBQRBKKCSRBwYBeN6ISZwmAQTZoSSg-5-Dp4tAUFrkRAAH0Qr3en2+iAAvCMxpNpo45rpmYg0DZ2WQuTyDnyvuMhSKJlMrK5JX8oABPYZoEkyp6C82IADkjXIlsQADJ7YjDcbSbKhRbrY47Y6zFFLDNyNRWexZSUDUaTayPcKva4fU6I66ZTYY1abWQE37qtqWHYQynwy6o2bPRms0mo6nzbHyw6nVVornyAkC2HdJW3aXa97687I27q2Xe77GwG6lQ2zYSgDgaCIdDYfDU0i8Yc6Yw4Hi0ZicXiAJIyyFG6DIPFQOAMplSFlxeWK3kfVXq0aa8U6krStC1e-cx-8tVhVfMVmzIXUhC-OIAEEDnBWooRgSQ8WFTsZTiNM4xYLN+2Tb8MPLXQ0Gg2D4MhRC8VHcwm0DShJziUNahKIiYPGOCEKQvsxxqVx83omVGMI4jWNI8jOKo8d4johimKEtiyI4yj-W4lh6iI6TtCAA)

Typescript Nightly produced more weird results: [Results in TSNext](https://www.typescriptlang.org/play?ts=4.4.0-dev.20210727#code/GYVwdgxgLglg9mABDASgUwmmA3NA5EAWwCM0AnARgAoBDALkTCNLIBpFiGBnKMmMAOYBKRAG8AUIimIICLnAA2aAHQK4A2ogDUHIQG5xAX3GhIsBMnSYc+ZuQBMtBkxLl2nRDz6CRE6TLlFFTUNGm1dA2NTaHgkVAwsXAJXMgBmJ0Y7Ng5uXn5hMUlpWTB5JVV1TR1ifSMTcBiLeOskrIAWDJcWd1zvAr9iwPKQqoi6+rNYxAAHMjgAExBMZJYqXyKpMjQoEDIkAFkaKAALZTIaMHm4QjXEAD5EAAZlAFZEAH5ECkQGcHm0YD8NDzSLicQAegAVJDJJDEABRAAeNEI0yUX0QJyOiAA7nAyABrLjuEBQRBqKBcRBwYABf6IRZomAQI5oLiw8Fgko8RBoCjOLKIAA+iD+AKB80QAF4ZnNFsssmsDNyyWh7AKUsLRZdxWBgdLZQslrYUkqwVAAJ7TNA03nfKUOxAAci65CdiAAZB7MVabbS1dLHS6su6vZYEjYVuRqHz2Grapbrba+YGZcGUqHvYm-bz7Knna6yJnwy0TSxHLHcwnfcn7UHC8Xs8m8w60w3Pd7moky+R0pX4wYm-6622Qx2fUn-S362Ow13I+0qP37LUIdDYQjkaj0XmsWSrmzGHAyXjCSSyQBJXmI63QZBkqBwDlcuSq1IalhasWAvWSmWzI0FVNWoVV5NoP3IL8dR-fV-zlY0ozIM1xFAtBUgAQUueE2iRGAeRlIdeVSfMnS8fJiwnHM0DaEiyMEJ0DDQzD5mw3CeTnKxu0QmNUjjNpaiYrCcMRPCyQ4iNWhSCteLAgSMKEtixM7TiFxSPsZOouTmNYkT2OUiSezIDo0L4-QgA) (Note the `string` and `number` should give error, but they don't)

### 💻 Code

```ts
function iReceiveNumber1(a: number, b: number) {
    console.log(a + b);
}


function produceNumber(): number | undefined {
    return Math.random() > 0.5 ? 1 : undefined;
}

/**
 * Example 1 that works, but lots of code duplicates
 */

const e1: number | undefined = produceNumber();
const e2: number | undefined = produceNumber();

typeof e1 === 'number' && typeof e2 === 'number' && iReceiveNumber1(e1, e2);

/**
 * Example 2 that does not work, but I expect it to
 */

const e3: number | undefined = produceNumber();
const e4: number | undefined = produceNumber();

const e3AndE4Exist = typeof e3 === 'number' &&  typeof e4 === 'number';
e3AndE4Exist && iReceiveNumber1(e3, e4);

// I receive error in above line, but I did not expect error
```

### 🙁 Actual behavior

I receive error that says Type 'undefined' is not assignable to type `number`. Although I checked the existence in a variable (`e3AndE4Exist`

### 🙂 Expected behavior

I expected TypeScript to automatically infer that the type is not `undefined` as I checked it in `e3AndE4Exist` variable right before I call the `iReceiveNumber` function.

I also had hard  times opening Typescript Playground, some requests failed. I don't know if that is my connection problem or not.

**Comment by MartinJohns** [comment]
This already works in TypeScript 4.4.0 due to #44730.

**Comment by RyanCavanaugh** [comment]
> Typescript Nightly produced more weird results: Results in TSNext

I don't see anything weird in the nightly build, can you explain what you find to be wrong?

**Comment by aghArdeshir** [comment]
I'm sorry. I checked again today, and in nightly version everything was still weird. I tried a simple `const a: number = 'a'` and again, I did not see any error. I think there is something wrong with my connection. By the way all requests from `https://typescript.azureedge.net/cdn/4.4.0-dev.20210727/...` fail.

I tried with version `4.4.0` locally and the issue is solved and everything was fine. Sorry for the inconvenience. :+1: 

I think this issue can be closed
